[WebKit] Enable InstallAPI for macOS/tvOS/watchOS

emw-apple opened the pull request:
#### 110ccc3202f36448008342c92ce503820c473a8c
<pre>
[WebKit] Enable InstallAPI for macOS/tvOS/watchOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=257560">https://bugs.webkit.org/show_bug.cgi?id=257560</a>
rdar://103361403

Reviewed by Alexey Proskuryakov.

Straightforward xcconfig and header changes to make WebKit and
WebKitLegacy use InstallAPI on all platforms. To avoid increasing the
potential for build failures due to stale headers, only introduce new
use of InstallAPI when TAPI_USE_SRCROOT is enabled.

* Source/WebKit/Configurations/Base.xcconfig: Remove platform conditions
  from SUPPORTS_TEXT_BASED_API.
* Source/WebKit/Configurations/WebKit.xcconfig: Remove platform
  conditions from OTHER_TAPI_FLAGS.
* Source/WebKit/Platform/ExtraPublicSymbolsForTAPI.h:
  $ld$install_name$os declarations are iOS-only. Add conditional guards.
* Source/WebKit/UIProcess/API/C/WKOpenPanelResultListener.h:
  WKOpenPanelResultListenerChooseMediaFiles is only implemented on iOS.
  Wrap it in a conditional guard.
* Source/WebKit/Shared/API/Cocoa/WebKit.m: Drive-by fixup: Change the
  ifdef condition to only emit versions-specific install names on iOS.

* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Add
  reference to ExtraPrivateSymbolsForTAPI.h
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig: Remove
  platform conditions from SUPPORTS_TEXT_BASED_API, TAPI_VERIFY_MODE,
  and OTHER_TAPI_FLAGS.
* Source/WebKitLegacy/mac/ExtraPrivateSymbolsForTAPI.h: Added, declares
  `WebCoreScrollView` which is implemented for backwards compatibility
  but removed from API/SPI headers.
* Source/WebKitLegacy/mac/Misc/WebIconDatabase.h:
  Declare WebIconDatabaseDidRemoveAllIconsNotification.
  Delete WebIconDatabaseEnabledDefaultsKey; it&apos;s not used anywhere and
  is not in any export lists.
* Source/WebKitLegacy/mac/Misc/WebIconDatabase.mm: Ditto.
* Source/WebKitLegacy/mac/Panels/WebPanelAuthenticationHandler.h: Move
  WebPanelAuthenticationHandler ivars to the .m file; they are neither
  exported nor used outside the implementation.
* Source/WebKitLegacy/mac/Panels/WebPanelAuthenticationHandler.m: See
  above.
* Source/WebKitLegacy/mac/WebCoreSupport/WebJavaScriptTextInputPanel.h:
  Ditto for WebJavaScriptTextInputPanel&apos;s ivars.
* Source/WebKitLegacy/mac/WebCoreSupport/WebJavaScriptTextInputPanel.m:
  See above.
* Source/WebKitLegacy/mac/WebKit.mac.exp: WebUndefined is declared in a
  migrated header (WebScriptObject.h) and used by DumpRenderTree, so it
  should be reexported.
* Source/WebKitLegacy/mac/WebView/WebViewInternal.h: Move unexported key
  strings here from WebViewPrivate.h.
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h: See above.

Xcode 13.x does not respect file filters (e.g.
EXCLUDED_SOURCE_FILE_NAMES) in TAPI filelists (rdar://92126006), so
raise the target version for TAPI_USE_SRCROOT mode to 14.0. This still
provides macOS and iOS coverage with SRCROOT mode in EWS. Also, the
feature flag can be passed as a defaults override, not an environment
variable, so do that instead.

* Tools/Scripts/build-webkit:
* Tools/Scripts/webkitdirs.pm:
(XcodeOptions):

Canonical link: <a href="https://commits.webkit.org/264993@main">https://commits.webkit.org/264993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcfc007cd70ce6984e6c41ed7c86749de5320e70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9371 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9647 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9878 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11032 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9245 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9380 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9609 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12130 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10438 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11189 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7735 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8549 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8037 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8832 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8699 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12042 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/8970 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9194 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7487 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9546 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8407 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2317 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2270 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12631 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9789 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8956 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2400 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/Platform/ExtraPublicSymbolsForTAPI.h:
```diff
@@ -25,6 +25,10 @@
 
+#include <wtf/Platform.h>
+
+#if PLATFORM(IOS_FAMILY) && !PLATFORM(WATCHOS) && !PLATFORM(APPLETV)
```
I think that this should be `PLATFORM(IOS) || PLATFORM(IOS_SIMULATOR)`. Doing it by exclusion is not forward looking in this case, but also, the current condition doesn't exclude macCatalyst.

emw-apple replied:
Makes sense. This is copied from the implementation site, so I'll change it in both places: https://github.com/WebKit/WebKit/blob/580bf6c28d2b381a9a51fb66f54c445bc51d967d/Source/WebKit/Shared/API/Cocoa/WebKit.m#L28

From my reading, this change will exclude Catalyst and the just-added Vision platform. I assume no one has noticed that these have been declared for either platform because we only declare version-specific install names for `4.3..7.1`, and those targets have not had OS versions in that range.

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -1227,8 +1227,9 @@ sub XcodeOptions
     push @options, "SDKROOT=$xcodeSDK" if $xcodeSDK;
-    if (xcodeVersion() lt "15.0") {
-        push @options, "TAPI_USE_SRCROOT=YES" if $ENV{UseSRCROOTSupportForTAPI};
+    if (xcodeVersion() ge "14.0" && xcodeVersion() lt "15.0") {
```
Is checking for Xcode version right here, or do we need to check toolchain?

emw-apple replied:
Xcode alone should be fine – what matters here is whether XCBuild is new enough to recognize `TAPI_USE_SRCROOT` without the defaults flag flipped.

emw-apple committed:
[WebKit] Enable InstallAPI for macOS/tvOS/watchOS
https://bugs.webkit.org/show_bug.cgi?id=257560
rdar://103361403

Reviewed by Alexey Proskuryakov.

Straightforward xcconfig and header changes to make WebKit and
WebKitLegacy use InstallAPI on all platforms. To avoid increasing the
potential for build failures due to stale headers, only introduce new
use of InstallAPI when TAPI_USE_SRCROOT is enabled.

* Source/WebKit/Configurations/Base.xcconfig: Remove platform conditions
  from SUPPORTS_TEXT_BASED_API.
* Source/WebKit/Configurations/WebKit.xcconfig: Remove platform
  conditions from OTHER_TAPI_FLAGS.
* Source/WebKit/Platform/ExtraPublicSymbolsForTAPI.h:
  $ld$install_name$os declarations are iOS-only. Add conditional guards.
* Source/WebKit/UIProcess/API/C/WKOpenPanelResultListener.h:
  WKOpenPanelResultListenerChooseMediaFiles is only implemented on iOS.
  Wrap it in a conditional guard.
* Source/WebKit/Shared/API/Cocoa/WebKit.m: Drive-by fixup: Change the
  ifdef condition to only emit versions-specific install names on iOS.

* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Add
  reference to ExtraPrivateSymbolsForTAPI.h
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig: Remove
  platform conditions from SUPPORTS_TEXT_BASED_API, TAPI_VERIFY_MODE,
  and OTHER_TAPI_FLAGS.
* Source/WebKitLegacy/mac/ExtraPrivateSymbolsForTAPI.h: Added, declares
  `WebCoreScrollView` which is implemented for backwards compatibility
  but removed from API/SPI headers.
* Source/WebKitLegacy/mac/Misc/WebIconDatabase.h:
  Declare WebIconDatabaseDidRemoveAllIconsNotification.
  Delete WebIconDatabaseEnabledDefaultsKey; it's not used anywhere and
  is not in any export lists.
* Source/WebKitLegacy/mac/Misc/WebIconDatabase.mm: Ditto.
* Source/WebKitLegacy/mac/Panels/WebPanelAuthenticationHandler.h: Move
  WebPanelAuthenticationHandler ivars to the .m file; they are neither
  exported nor used outside the implementation.
* Source/WebKitLegacy/mac/Panels/WebPanelAuthenticationHandler.m: See
  above.
* Source/WebKitLegacy/mac/WebCoreSupport/WebJavaScriptTextInputPanel.h:
  Ditto for WebJavaScriptTextInputPanel's ivars.
* Source/WebKitLegacy/mac/WebCoreSupport/WebJavaScriptTextInputPanel.m:
  See above.
* Source/WebKitLegacy/mac/WebKit.mac.exp: WebUndefined is declared in a
  migrated header (WebScriptObject.h) and used by DumpRenderTree, so it
  should be reexported.
* Source/WebKitLegacy/mac/WebView/WebViewInternal.h: Move unexported key
  strings here from WebViewPrivate.h.
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h: See above.

Xcode 13.x does not respect file filters (e.g.
EXCLUDED_SOURCE_FILE_NAMES) in TAPI filelists (rdar://92126006), so
raise the target version for TAPI_USE_SRCROOT mode to 14.0. This still
provides macOS and iOS coverage with SRCROOT mode in EWS. Also, the
feature flag can be passed as a defaults override, not an environment
variable, so do that instead.

* Tools/Scripts/build-webkit:
* Tools/Scripts/webkitdirs.pm:
(XcodeOptions):

Canonical link: https://commits.webkit.org/264993@main
